src/audio/codec_adapter/codec/dts.c

```diff
@@ -0,0 +1,365 @@
+
+	comp_dbg(dev, "dts_codec_init_process() done");
+
+	return ret;
+}
+
+int dts_codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	DtsSofInterfaceResult dts_result;
+	unsigned int bytes_processed = 0;
+
+	comp_dbg(dev, "dts_codec_process() start");
+
+	dts_result = dtsSofInterfaceProcess(codec->private, &bytes_processed);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	codec->cpd.consumed = !ret ? bytes_processed : 0;
+	codec->cpd.produced = !ret ? bytes_processed : 0;
+
+	if (ret)
+		comp_err(dev, "dts_codec_process() failed %d %d", ret, dts_result);
+
+	comp_dbg(dev, "dts_codec_process() done");
+
+	return ret;
+}
+
+int dts_codec_apply_config(struct comp_dev *dev)
+{
+	int ret = 0;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct codec_config *config;
+	struct codec_param *param;
+	uint32_t config_header_size;
+	uint32_t config_data_size;
+	uint32_t param_header_size;
+	uint32_t param_data_size;
+	uint32_t i;
+	uint32_t param_number = 0;
+	DtsSofInterfaceResult dts_result;
+
+	comp_dbg(dev, "dts_codec_apply_config() start");
+
+	config = &codec->r_cfg;
+
+	/* Check that config->data isn't invalid and has size greater than 0 */
+	config_header_size = sizeof(config->size) + sizeof(config->avail);
+	if (config->size < config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() config->data is invalid");
+		return 0;
+	} else if (config->size == config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() size of config->data is 0");
+		return 0;
+	}
+
+	/* Calculate size of config->data */
+	config_data_size = config->size - config_header_size;
+
+	/* Check that config->data is not greater than the max expected for DTS data */
+	if (config_data_size > MAX_EXPECTED_DTS_CONFIG_DATA_SIZE) {
+		comp_dbg(dev,
+			"dts_codec_apply_config() size of config->data is larger than max for DTS data");
+		return 0;
+	}
+
+	/* Allow for multiple codec_params to be packed into the data pointed to by config
+	 */
+	for (i = 0; i < config_data_size; param_number++) {
+		param = (struct codec_param *)((char *)config->data + i);
+		param_header_size = sizeof(param->id) + sizeof(param->size);
+
+		/* If param->size is less than param_header_size, then this param is not valid */
+		if (param->size < param_header_size) {
+			comp_dbg(dev, "dts_codec_apply_config() param is invalid");
+			return 0;
+		}
+
+		/* Only process param->data if it has size greater than 0 */
+		if (param->size > param_header_size) {
+			/* Calculate size of param->data */
+			param_data_size = param->size - param_header_size;
+
+			if (param_data_size) {
+				dts_result = dtsSofInterfaceApplyConfig(codec->private, param->id,
+					param->data, param_data_size);
+				ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+				if (ret) {
+					comp_err(dev,
+						"dts_codec_apply_config() dtsSofInterfaceApplyConfig failed %d",
+						dts_result);
+					break;
```

Comment:
Can we recover from this? If not we should exit not continue (prevent future modifications adding code on an error path by accident)